website_canonical_url/models/website.py

```diff
@@ -1,35 +1,73 @@
+# Copyright initOS GmbH 2016
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
-from openerp import models, api
-from openerp.http import request
-from urlparse import urlparse, urlunparse
+from odoo import models, api, fields, exceptions, _
+from odoo.http import request
+from urlparse import urlparse, urljoin
 
 
-class website(models.Model):
+class Website(models.Model):
     _inherit = 'website'
 
+    canonical_domain = fields.Char(
+        help='Canonical domain is used to build unique canonical URLs '
+             'to make SEO happy.'
+    )
+
+    @api.constrains('canonical_domain')
+    def _check_canonical_domain(self):
+        domain = self.canonical_domain
+        if domain and not urlparse(domain).scheme:
+            raise exceptions.ValidationError(_(
+                'Canonical domain must contain protocol `http(s)://`'
+            ))
+
     @api.multi
     def get_canonical_url(self, req=None):
-        canonical_url = None
-        if req is None:  # pragma: no cover
-            req = request
-        if req and req.httprequest.path:
-            lang = self.env.lang
-            if lang == request.website.default_lang_code:
-                canonical_url = req.httprequest.path
-            else:
-                url_parts = urlparse(req.httprequest.path)
-                url_parts = list(url_parts)
-                # change the path of the url
-                url_parts[2] = lang + url_parts[2]
-                canonical_url = urlunparse(url_parts)
-        # Special case for rerouted requests to root path
-        try:
-            if (canonical_url.startswith("/page/") and
-                    self.menu_id.child_id[0].url == canonical_url):
-                canonical_url = "/"
-        except:  # pragma: no cover
-            pass
+        return urljoin(
+            self._get_canonical_domain(),
+            self._get_canonical_relative_url(req=req)
+        )
+
+    @api.multi
+    def _get_canonical_domain(self):
+        self.ensure_one()
+        if self.canonical_domain:
+            return self.canonical_domain
+        params = self.env['ir.config_parameter'].sudo()
+        return params.get_param('web.base.url')
+
+    @api.multi
+    def _get_canonical_relative_url(self, req=None):
+        req = req or request
+        parsed = urlparse(req.httprequest.path)
+        # here we get the full path and qstring args stripped out already
+        canonical_url = parsed.path
+        # build lang path if request lang does not match default lang
+        lang_path = ''
+        if req.lang != req.website.default_lang_code:
+            # eg: language is `it_IT` but main lang is `en_US`
+            lang_path = '/%s/' % req.lang
+        # requests might be redirected or rerouted,
+        # when this happens here we don't have the lang in the path anymore
+        if lang_path and not canonical_url.startswith(lang_path):
+            canonical_url = urljoin(lang_path, canonical_url.lstrip('/'))
```

Comment:
Oh come on, not again... :man_facepalming: 

Please, friend, I really don't wanna be rude, really :sweat_smile: :v: , but requests *always* get rerouted. Don't complicate things, it's extremely simple:

```py
if "/" in getattr(req, "rerouting", []) or req.httprequest.path == self.menu_id.child_id[:1].url:  # <-- This can go into `_is_root_page()` if you want, and you can use the more complex algorithm you linked before also if you want
    return "/"
if req.lang != req.website.default_lang_code:
    return "/%s%s" % (req.lang, req.httprequest.path)
return req.httprequest.path  # This already strips out hostname and query string
```

Please KISS.

Reply:
Oh come on, not again: is this "KISSy" solution more clear? I don't think so :stuck_out_tongue: 

What you have more in my solution is:

1. parsing the path -> I can remove it if that's stripped already [done]
2. check if the lang is there in the path, which I consider more safe anyway (WOW, what a non KISSy thing!!).
Also, building the lang path IMHO is better when you have to test the menu item: users can copy paste URLs w/ the lang in the URL (yes, they can do it :sweat_smile:) so in that case my check is more defensive
3. I moved the check for the root menu to make things more readable and hackable in case

Reply:
> check if the lang is there in the path, which I consider more safe anyway (WOW, what a non KISSy thing!!).

Yes, it's not KISSy. Simply because the lang will **never** be in the path.

> users can copy paste URLs w/ the lang in the URL

Imagine this case:

- Website default lang is en_US
- Website alt lang is es_ES
- User sets the 1st menu item URL to `/es_ES/`
- Somebody clicks in the main link

The result of this is an endless 302 redirection loop.

So, if users copy a lang-specific link in a menu item, the canonical url will only be one of the problems it will produce. There will be more. That's simply a **not supported** scenario.

BTW, while editing a website, you always edit in the default language, because other ones only have translation mode, so it's not much likely that will happen.

So, basically, any scenario that justifies this code is simply impossible. Remove it please.